Add allocator for node based containers

martinus opened the pull request:
jamesob has benchmarked my rebased branch (see https://github.com/bitcoin/bitcoin/pull/16801#issuecomment-895348607) concerning the allocator for node based containers, and has seen a 9% speedup along with 8% reduction in memory. That's why I had another good look at my previously closed PR https://github.com/bitcoin/bitcoin/pull/16801 and updated the code. I've updated the code quite a bit, mostly trying to simplify it and improve correctness:

* Renamed classes to be more in line with naming used for the`std::pmr` allocators (see https://en.cppreference.com/w/cpp/memory/polymorphic_allocator)
* Simplified `Allocator` thanks to being able to use C++17 behavior (`std::allocator_traits`)
* Simpler allocation: only allocate blocks of the same size and just keep them in a `std::vector`.
* Only access memory when actually needed. E.g. Linux uses optimistic memory allocation, so memory pages from malloc are only actually made available to the program when accessed
* Correctly handle alignment for the inplace linked list.

jamesob it would be great if you could run your benchmarks again with the updated code. 

edit laanwj: removed @ signs and html comments from message

practicalswift commented:
Concept ACK

The numbers look very promising!

Seems like this could be one of those rare optimization opportunities actually worth pursuing :)

JeremyRubin commented on src/support/allocators/node_allocator.h:
```diff
@@ -0,0 +1,356 @@
+ * - Memory that's been used for nodes is always put back into a freelist and never given back to the system. Memory
+ *   is only freed when the MemoryResource is destructed.
+ *
+ * - The freelist is a simple first-in-last-out linked list, it doesn't reorder elements. So freeing and malloc'ing again
```
last in first out?

JeremyRubin commented on src/support/allocators/node_allocator.h:
```diff
@@ -0,0 +1,356 @@
+ * unfortunately no way to determine the size of nodes that we want to optimize for in advance, MemoryResource uses
+ * a simple heuristic: We assume the first call to Allocate with 1 element is for the node, and upon that first call the
+ * MemoryResource is configured to use that as it's chunk size. Note that due to this heuristic we cannot implement an
+ * `std::pmr::memory_resource` because it does not gie the information about the number of elements.
```
gie -> give

JeremyRubin commented:
I'm curious what the impact would be on the performance if you allocated N chunks at a time, always, to fill a page.

E.g., if you want to allocate one node, always allocate something like `aligned_alloc(sysconf(_SC_PAGESIZE), size)` and divide it up into a bunch of pointers for the free list.

Cool properties of doing so: Better cache alignment on items added around the same time, fewer things to keep in the allocated chunks list, if you were to sort the list by pointer (perhaps lazily, when we're not doing anything else) you would get something with good cache alignment again.

martinus commented:
> I'm curious what the impact would be on the performance if you allocated N chunks at a time, always, to fill a page.

I've run the benchmark with `::aligned_alloc(sysconf(_SC_PAGESIZE), ...);` and only allocating a single page, and at least in the benchmark there's no difference:

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|              140.95 |        7,094,953.03 |    0.2% |      0.08 | `NodeAllocator_StdUnorderedMap`
|              115.57 |        8,652,942.75 |    0.1% |      0.06 | `NodeAllocator_StdUnorderedMapWithNodeAllocator` ::operator new
|              115.33 |        8,670,873.08 |    0.2% |      0.06 | `NodeAllocator_StdUnorderedMapWithNodeAllocator`aligned_alloc

JeremyRubin commented:
Do you have specific code for that? The changes sounded a bit more complex than just aligning the allocator. Did you divide the allocation into N free list chunks? Or just over-allocate?

martinus commented:
> Do you have specific code for that? The changes sounded a bit more complex than just aligning the allocator. Did you divide the allocation into N free list chunks? Or just over-allocate?

`MemoryResource::AllocateNewBlock` is the only place where I actually malloc a new block that's divided into the chunks. Here I've replaced the `::operator new(num_chunks* m_chunk_size_bytes)` with `::alligned_alloc(sysconf(_SC_PAGESIZE), num_chunks* m_chunk_size_bytes)`. And use `free` in `~MemoryResource`.
I also changed the default allocation size `m_block_size_bytes` from 262144 to `sysconf(_SC_PAGESIZE)`. 

So for node size of 104 bytes and page size 4096, this should allocate a page-alligned block with 39*104=4056 bytes.

jamesob commented:
Some fresh benches in (high dbcache). Continuing to see fairly substantial time/memory improvements with this change.

![ibd local range 500000 540000](https://user-images.githubusercontent.com/73197/129577300-d73f42aa-8598-4418-9375-2a7c4521701a.png)

#### commands index
|          bench name           |                                                                                                     command                                                                                                      |
|-------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ibd.local.range.500000.540000 | `bitcoind -dbcache=10000 -debug=coindb -debug=bench -listen=0 -connect=0 -addnode=127.0.0.1:8888 -prune=9999999 -printtoconsole=0 -assumevalid=000000000000000000176c192f42ad13ab159fdb20198b87e7ba3c001e47b876` |


#### #22702 vs. $mergebase (absolute)
|                  bench name                   |  x  |           #22702           |        $mergebase         |
|-----------------------------------------------|----:|----------------------------|---------------------------|
| ibd.local.range.500000.540000.total_secs      |   3 | 2888.4055 (± 24.7775)      | 3131.5871 (± 4.8045)      |
| ibd.local.range.500000.540000.peak_rss_KiB    |   3 | 5886729.3333 (± 3815.6315) | 6352470.6667 (± 979.1552) |
| ibd.local.range.500000.540000.cpu_kernel_secs |   3 | 263.5433 (± 1.9605)        | 269.2233 (± 2.6299)       |
| ibd.local.range.500000.540000.cpu_user_secs   |   3 | 12852.1300 (± 15.0661)     | 13134.5300 (± 2.9587)     |


#### #22702 vs. $mergebase (relative)
|                  bench name                   |  x  | #22702 | $mergebase |
|-----------------------------------------------|----:|-------:|-----------:|
| ibd.local.range.500000.540000.total_secs      |   3 |      1 |      1.084 |
| ibd.local.range.500000.540000.peak_rss_KiB    |   3 |      1 |      1.079 |
| ibd.local.range.500000.540000.cpu_kernel_secs |   3 |      1 |      1.022 |
| ibd.local.range.500000.540000.cpu_user_secs   |   3 |      1 |      1.022 |

martinus commented:
Thanks for the benchmark! Good to see its 8.4% faster. I think the only reason the memory is lower is because the `memusage::DynamicUsage` now overestimates the memory requirements of the `std::unordered_map` that uses the node_allocator. I guess I should correct that. Then the memory usage should stay roughly the same (as it should, that's what `-dbcache` is for), but the number of transactions that can be cached will increase.

JeremyRubin commented:
> `MemoryResource::AllocateNewBlock` is the only place where I actually malloc a new block that's divided into the chunks. Here I've replaced the `::operator new(num_chunks* m_chunk_size_bytes)` with `::alligned_alloc(sysconf(_SC_PAGESIZE), num_chunks* m_chunk_size_bytes)`. And use `free` in `~MemoryResource`.
> I also changed the default allocation size `m_block_size_bytes` from 262144 to `sysconf(_SC_PAGESIZE)`.

hmm yeah it's interesting to think through the alignment issues. I think I understand the code a bit better now so it makes sense that most of you allocations are already aligned. One thing that strikes me is that it seems that you have these 104 byte chunks (13 uint64_t s). You might get better properties if you pad them out to the nearest 32/64/128 bytes (memory overhead 24 bytes :( ), because then you'll never have nodes across cache lines. Memory usage is a bit worse though, but cache perfomance may be better this way? https://www.usenix.org/legacy/publications/library/proceedings/als00/2000papers/papers/full_papers/sears/sears_html/index.html for reference.

JeremyRubin commented:
might also be interesting to bump the CScript prevector size from 28 to 35 (7 bytes). This would help with the above as well because if we're 104/128 byte aligned we have the capacity & then we'd also save a prevector allocation (which means the vector costs 25 bytes + 34-35 bytes) for a non p2sh segwit v0-v1 output, and also means even more indirection.

This might have system impacts very broadly because prevector is a leaky optimization sadly.

jonatack commented on src/support/allocators/node_allocator.h:
```diff
@@ -0,0 +1,356 @@
+     * Copying a memory resource is not allowed, it is an immobile object.
+     */
+    MemoryResource(const MemoryResource&) = delete;
+    MemoryResource& operator=(const MemoryResource&) = delete;
```
Perhaps rule of 5 (https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-five, https://www.stroustrup.com/C++11FAQ.html#default)
```diff 
     /**
-     * Copying a memory resource is not allowed, it is an immobile object.
+     * Copying/moving a memory resource is not allowed; it is an immobile object.
      */
     MemoryResource(const MemoryResource&) = delete;
     MemoryResource& operator=(const MemoryResource&) = delete;
+    MemoryResource(MemoryResource&&) = delete;
+    MemoryResource& operator=(MemoryResource&&) = delete;
```

jonatack reviewed (commented):
Concept/Approach ACK. Interesting work! Some initial review feedback at https://github.com/jonatack/bitcoin/commit/df355206323e61603b181d426f4977f026559784

0xB10C commented:
I've benchmarked the performance of `CChainState::ConnectBlock` between cde8a991f525b72d3a7ac76e0c83aaa611169f22 
 (PR)  and the mergebase 803ef70fd9f65ef800567ff9456fac525bc3e3c2 (MB) using the [`validation::connect_block`](https://github.com/bitcoin/bitcoin/blob/master/doc/tracing.md#tracepoint-validationblock_connected) tracepoint. The tracepoint reports the time it took to connect each block. This means my results don't include the time it took to download and, AFAIK, not the time it took to persist the block to disk. I didn't look at memory usage. 

I've oriented myself on @jamesob's parameters and synced the blocks 500.000 to 540.000 from a localhost peer three times for the PR and the MB. I used the default dbcache (shouldn't matter for my measurements as cache flushes don't happen while inside `CChainState::ConnectBlock`). I've used an idle workstation with an i5 6500 and an NVME drive. Room and hardware temperature between runs changed which could have impacted performance. A `bpftrace` script to hook into the tracepoint can be found [here](https://gist.github.com/0xB10C/7f9010d44f1d97fbf1c83ec7f98155fd).

Total time spent in  `CChainState::ConnectBlock` per run for PR and MB: 
| run | 1         | 2     | 3     | 
|-----|-----------|-------|-------|
| PR  | **1427s** | 1487s | 1486s | 
| MB  | **1577s** | 1670s | 1661s |

![image](https://user-images.githubusercontent.com/19157360/129805353-5b48c0c1-1184-49c2-800c-196f1eed89c8.png)
(PR run 2 and run 3 are overlapping quite a bit and indistinguishably here)

My results confirm that less time is spent in `CChainState::ConnectBlock` with PR compared to MB. A performance improvement between 6% - 10% is likely. 


To rule out temperature effects on performance, these measurements could be redone with a properly cooled server/workstation.

martinus commented:
@JeremyRubin I actually tried to bump it up to 128 byte to see what that does, but I didn't see any major difference. The benchmark got maybe 1% slower, but maybe the benchmark is also not very realistic. I also tried to reduce the size by 8 byte (which can be done by making use of the `Coin`'s padding in `CCoinsCacheEntry`, I might open anothe PR for that), and it got about ~1% faster. I think though that even with no performance benefit the node_allocator is worthwhile because it has lower memory overhead, so we can cache more entries with the same `-dbcache` setting.

@jonatack Thanks for the review! I'm not sure though how I should best incorporate your changes? Should I copy them and add a `Co-authored-by` in the commit message? (I've never done that)

@0xB10C Thanks for the benchmark! You might get a bit more stable result with [pyperf](https://pyperf.readthedocs.io/en/latest/): I always use `sudo pyperf system tune` before I run a benchmark which does a few things to make benchmarks much more stable (e.g. locks the CPU to a fixed frequency, disables turbo boost)

jonatack commented:
> @jonatack Thanks for the review! I'm not sure though how I should best incorporate your changes? Should I copy them and add a `Co-authored-by` in the commit message? (I've never done that)

Thanks! I'd say squash in the changes you want to keep. You can add an empty line followed by `Co-authored-by: Jon Atack <jon@atack.com>` to the bottom of a commit message if you like, but it was just a regular average review :)

jamesob commented:
Bench results with more modest dbcache (800): 6% speedup, 6.7% less memory usage.

|          bench name           |                                                                                                    command                                                                                                     |
|-------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ibd.local.range.500000.540000 | `bitcoind -dbcache=800 -debug=coindb -debug=bench -listen=0 -connect=0 -addnode=127.0.0.1:8888 -prune=9999999 -printtoconsole=0 -assumevalid=000000000000000000176c192f42ad13ab159fdb20198b87e7ba3c001e47b876` |

|                  bench name                   |  x  |           #22702            |         $mergebase          |
|-----------------------------------------------|----:|-----------------------------|-----------------------------|
| ibd.local.range.500000.540000.total_secs      |   3 | 3519.2667 (± 58.0440)       | 3752.9732 (± 54.8626)       |
| ibd.local.range.500000.540000.peak_rss_KiB    |   3 | 2282000.0000 (± 45539.0428) | 2443188.0000 (± 22211.2803) |
| ibd.local.range.500000.540000.cpu_kernel_secs |   3 | 491.2067 (± 4.5352)         | 493.7567 (± 13.1047)        |
| ibd.local.range.500000.540000.cpu_user_secs   |   3 | 13717.5467 (± 15.0752)      | 13957.7867 (± 18.7773)      |

|                  bench name                   |  x  | #22702 | $mergebase |
|-----------------------------------------------|----:|-------:|-----------:|
| ibd.local.range.500000.540000.total_secs      |   3 |      1 |      1.066 |
| ibd.local.range.500000.540000.peak_rss_KiB    |   3 |      1 |      1.071 |
| ibd.local.range.500000.540000.cpu_kernel_secs |   3 |      1 |      1.005 |
| ibd.local.range.500000.540000.cpu_user_secs   |   3 |      1 |      1.018 |

martinus commented:
I've now updated the code with correct memory estimation, and benchmarked it. I had a fully synchronized node, and ran `-reindex-chainstate` from block 0 to 690000. All done on a Intel i7 8700, locked to 3200 MHz.

```sh
/usr/bin/time -v bitcoind -datadir=/run/media/martinus/big/bitcoin/db -dbcache=5000 -assumevalid=00000000000000000002a23d6df20eecec15b21d32c75833cce28f113de888b7 -reindex-chainstate -printtoconsole=0 -stopatheight=690000
```
This ran quite a bit faster, 20.8%, with practically the same memory usage. Here some interesting data from `/usr/bin/time`:

| what | mergebase | #22702 | Change 
---|---:|---:|---:
Elapsed (wall clock) time (h:mm:ss or m:ss) | 4:06:05 | 3:14:49 | 20.8% lower
Maximum resident set size (kbytes) | 6837496 | 6856104 | 0.27% more
Minor (reclaiming a frame) page faults | 101689566 | 70194069 | 31% fewer
File system inputs | 693891464 | 687479120 | 0.9% fewer
File system outputs | 239830024 | 188574064 | 21.4% fewer

Here are some graphs that I created from parsing the `debug.log` file:

![Progress in Million Transactions over Time(1)](https://user-images.githubusercontent.com/14386/130218881-c724daa3-f1ab-46e4-848f-a467fe875e4f.png)

![Size of Cache in MiB over Time](https://user-images.githubusercontent.com/14386/130218975-c04af0ce-d36b-4fd6-9a4d-64b43040650c.png)

`node_allocator` requires currently about 104 byte for one node in the hashmap (a bit more for house keeping but that's practically negigable), and using malloc requires 128 byte (16 byte overhead and 8 more byte for 16 byte alignment). So we can cache quite a bit more transactions in the same amount of memory:

![Size of Cache in Million tx over Time](https://user-images.githubusercontent.com/14386/130219000-5f9674cd-fffb-41c8-aa48-85ce51daca77.png)

Due to the different allocation behavior (it allocates one big chunk of memory right as the first Coin is added), I also had to update the checks in `validation_flush_tests.cpp`

sipa commented on src/support/allocators/node_allocator.h:
```diff
@@ -0,0 +1,399 @@
+    //! A contains all allocated blocks of memory, used to free the data in the destructor.
+    std::vector<void*> m_allocated_blocks{};
+
+    //! The pool's size for the memory blocks. First call to Allocate() determines the used size.
```
I think this comment is incorrect (this is the size of one chunk; blocks are a multiple of that).

sipa commented on src/support/allocators/node_allocator.h:
```diff
@@ -0,0 +1,399 @@
+        }
+
+        // peel off one chunk from the untouched memory
+        auto tmp = m_untouched_memory_iterator;
```
Perhaps add a comment here explaining that the `next` pointer of in-use elements actually doesn't matter until it's deallocated, so it doesn't need initialization here.

martinus commented on src/support/allocators/node_allocator.h:
```diff
@@ -0,0 +1,399 @@
+ * - Memory that's been used for nodes is always put back into a free list and never given back to the system. Memory
+ *   is only freed when the MemoryResource is destructed.
+ *
+ * - The free list is a simple last-in-first-out linked list, it doesn't reorder elements. So freeing and malloc'ing again
```
yes I wanted to clarify that the linked list's can become relatively random access pattern into the memory, which can be slower due to lots of cache misses. I'll fix the comment

martinus commented on src/support/allocators/node_allocator.h:
```diff
@@ -0,0 +1,399 @@
+    //! A single linked list of all data available in the pool. This list is used for allocations of single elements.
+    void* m_free_chunks = nullptr;
+
+    //! A contains all allocated blocks of memory, used to free the data in the destructor.
```
Should be just "Contains all allocated blocks of memory"

martinus commented:
Unfortunately Microsofts implementation of unordered_map behaves quite a bit different from libc++ and libstdc++, so the heuristic that I'm using to detect the node size doesn't work. I've added a WIP to the header until I've fixed this. Most likely I'll ditch the heuristic completely, and properly calculate the correct size. I'm already doing that in the tests anyways (except for windows)

jonatack reviewed (commented):
Currently-testing almost-ACK c6561fd394aab850ac2d8bc6889eaaa274c4abda

Learning more about allocators thanks to this patch and the links in your documentation.

A second round of review ideas at https://github.com/jonatack/bitcoin/commits/pr22702-review-2, feel free to pick/choose/ignore/squash in. I'm running a node built on that/this branch.

jonatack reviewed (commented):
ACK 952c37a31c39b7acdb5ab2634a354a1c548708a2 code review and tested on debian 5.10.46-4 (2021-08-03) running a (synced) mainnet node since 8 days so far

martinus commented on src/bench/node_allocator.cpp:
```diff
@@ -0,0 +1,52 @@
+
+    bench.epochIterations(5000 * 10).run([&] {
+        // modify hash a bit so we get a new entry in the map
+        ++p.n;
```
I think `p.n` cannot overflow here. It is initialized to 0, and since I've specified the number of iterations each measurement will have exactly 50000 iterations. I do this so I can be sure each measurement has exactly the same number of calls to `map.clear();`, which gives more stable benchmark results. Nanobench defaults to 11 measurements, so in total n will reach 550000.

0xB10C commented:
I think it would be better to not @ @jonatack in the commit message. IIRC he'll get a notification each time this commit is included in some bitcoin-fork.

martinus commented:
> I think it would be better to not @ @jonatack in the commit message. IIRC he'll get a notification each time this commit is included in some bitcoin-fork.

Ah right, it even says so in `CONTRIBUTING.md`. I'll change the commit message.

jonatack commented:
The @ is not bothersome until merge, so it can wait until need to retouch.

benthecarman commented:
tACK 952c37a31c39b7acdb5ab2634a354a1c548708a2

Did IDB on a testnet node, ubuntu 20.04

ryanofsky commented:
> @sipa @ryanofsky @TheBlueMatt: given you're some resident C++/systems experts, are there any concerns you have? Tests you'd like to see?

No specific concerns from me. It's just a PR adding a lot of dense code that will take time to read over. This seems like a very worthwhile change if it is speeding up IBD, and it seems less risky than other more invasive performance improvements.

Dumb question, just because the title of this PR is makes me immediately feel stupid: What is  a "node-based" container? Just any map container? Maybe a more motivating PR title would be "Implement node allocator for UTXO cache for faster IBD". Also it would be good if PR description described the new PR instead of just comparing it to previous PR, since probably some potential reviewers like me are not familiar with the previous PR.

ryanofsky commented:
Yeah I don't mean to quibble over the name. I'm sure it's appropriate. Just suggesting to make not make the term seem like assumed knowledge for reviewing the PR. Naively I would think "separately-allocated object" would apply to trees and lists but not hash tables so `{,unordered_}` part of this is specifically what's confusing me here. But I should just sit down and read `node_allocator.h`, since the code itself is nicely documented.

ryanofsky reviewed (approved):
Code review ACK 952c37a31c39b7acdb5ab2634a354a1c548708a2 for everything except benchmarks & tests, which I skimmed but didn't really look at yet.

Looks great! I left a lot of suggestions, but none are critical so please ignore them if they don't seem worth implementing. Some of them might make more sense to follow up later, anyway.

martinus committed:
Don't call ReallocateCache twice in  ResizeCoinsCaches

Removes the call to CoinsTip().ReallocateCache() in ResizeCoinsCaches because this is now done in CCoinsViewCache::Flush().

Also makes ReallocateCache private because it's not needed outside any more.

ryanofsky reviewed (commented):
This looks great and thanks for using some my earlier suggestions. Again feel free to ignore all suggestions from me, as I'm mostly writing to help understand things.

I mostly finished going through the review again, except I still didn't look deeply at test yet. I'm just posting comments mid-review now because they are piling up

I'd also recommend squashing all commits exception the validation `ReallocateCache` commit to avoid churn within PR and make review more straightforward for new reviewers.

ryanofsky reviewed (approved):
Code review ACK a91b6032f037ab1a26ac4baa761e809fbfeb1bdd. My suggestions are all minor and can be ignored or could be followed up later.

martinus commented:
Thanks again @ryanofsky, I'll of course look through your review closely again but I don't have much time right now, so it can take a while.

martinus commented on src/bench/node_allocator.cpp:
```diff
@@ -0,0 +1,52 @@
+template <typename Map>
+void BenchFillClearMap(benchmark::Bench& bench, Map& map)
+{
+    CMutableTransaction tx = CMutableTransaction();
+    tx.vin.resize(1);
+    tx.vin[0].scriptSig = CScript() << OP_2;
+    tx.vin[0].scriptWitness.stack.push_back({2});
+    tx.vout.resize(1);
+    tx.vout[0].scriptPubKey = CScript() << OP_2 << OP_EQUAL;
+    tx.vout[0].nValue = 10 * COIN;
```
right! I just copied that code without really thinking too much about it. I'll clean that up.

martinus replied:
Done in e2a5547f12e5da2718ba17ebac5d5aeb5a2e24d5

martinus commented on src/support/allocators/node_allocator.h:
```diff
@@ -0,0 +1,406 @@
+ * In-place linked list of the allocations, used for the free list.
+ */
+struct FreeList {
+    void* next;
```
Reading through the code again I now think that I actually have undefined behavior: I'm basically `reinterpret_cast`ing `T*` to `FreeList*`, and both types are completely unrelated, so according to the type aliasing rules accessing this is undefined behavior:

> Performing a class member access that designates a non-static data member or a non-static member function on a glvalue that does not actually designate an object of the appropriate type - such as one obtained through a reinterpret_cast - results in undefined behavior

See https://en.cppreference.com/w/cpp/language/reinterpret_cast

This can be fixed by using `memcpy`, and then I can also relax the alignment.

martinus replied:
I guess I didn't have enough coffee yet. I can't use reinterpret_cast, but since the `T` was already destroyed, I can use placement new in that memory to create the `FreeList`. No reinterpret_cast needed at all.

martinus replied:
`MemoryResource::Deallocate` now uses pacement new in e2a5547f12e5da2718ba17ebac5d5aeb5a2e24d5, and uses the returned pointer.

martinus commented on src/support/allocators/node_allocator.h:
```diff
@@ -0,0 +1,406 @@
+     * Creates the std::unordered_map container, and asserts that the specified memory_resource is
+     * correct.
+     */
+    [[nodiscard]] static ContainerType CreateContainer(MemoryResourceType* memory_resource)
```
I'd like to keep the `CreateContainer` here because otherwise constructing an unordered_map is quite a lot of non-obvious code, like first argument is always 0, then create a temporary hash of the correct typ, equals, and finally the Allocator.

martinus commented on src/support/allocators/node_allocator.h:
```diff
@@ -0,0 +1,406 @@
+class UnorderedMapFactory
+{
+public:
+    static constexpr size_t ALLOCATION_SIZE_BYTES = memusage::NodeSize<std::unordered_map<Key, Value, Hash, Equals>>::Value();
```
Thanks, and I agree that is the correct way to do it. I don't think it was a bug though, at least the worst thing that could happen was that it falls back to default `::operator new`, but for that there are unit tests too.

martinus replied:
Done in e2a5547f12e5da2718ba17ebac5d5aeb5a2e24d5

martinus commented on src/support/allocators/node_allocator.h:
```diff
@@ -0,0 +1,406 @@
+    using propagate_on_container_swap = std::true_type; // to avoid the undefined behavior
+
+    /**
+     * Move and swap have to propagate the allocator, so for consistency we do the same for copy
+     * assignment.
```
no real downside, it's just that because the allocator has a state it also has to be propagated.

martinus commented on src/test/node_allocator_tests.cpp:
```diff
@@ -0,0 +1,424 @@
+    m = Factory::CreateContainer(&mr);
+
+    // now we got everything free
+    BOOST_CHECK(node_allocator::MemoryResourceTester::NumFreeAllocations(mr) >
+                num_free_allocations + 50);
```
it sure can be made stricter :+1:

martinus replied:
stricter in e2a5547f12e5da2718ba17ebac5d5aeb5a2e24d5

martinus commented on src/test/node_allocator_tests.cpp:
```diff
@@ -0,0 +1,424 @@
+
+template <typename T>
+struct NotNoexceptHash {
+    size_t operator()(const T& x) const /* noexcept */ { return std::hash<T>{}(x); }
```
`noexcept(false)` and writing nothing is equivalent

martinus commented on src/memusage.h:
```diff
@@ -148,23 +154,65 @@ static inline size_t DynamicUsage(const std::shared_ptr<X>& p)
+        // because multiple maps could use the same MemoryResource.
+        return MallocUsage(sizeof(void*) * m.bucket_count());
+    } else {
+        auto node_size = NodeSize<std::unordered_map<Key, Value, Hash, Equals, Alloc>>::Value();
```
Split that code up into two methods in e2a5547f12e5da2718ba17ebac5d5aeb5a2e24d5

martinus commented on src/memusage.h:
```diff
@@ -148,23 +154,65 @@ static inline size_t DynamicUsage(const std::shared_ptr<X>& p)
 {
-    return MallocUsage(sizeof(unordered_node<std::pair<const X, Y> >)) * m.size() + MallocUsage(sizeof(void*) * m.bucket_count());
+    using ContainerType = std::unordered_map<Key, Value, Hash, Equals, Alloc>;
+    if constexpr (std::is_same_v<Alloc, node_allocator::Allocator<std::pair<const Key, Value>, memusage::NodeSize<ContainerType>::Value()>>) {
```
done in e2a5547f12e5da2718ba17ebac5d5aeb5a2e24d5

martinus commented on src/support/allocators/node_allocator.h:
```diff
@@ -0,0 +1,406 @@
+ * ## Behavior
+ *
+ * MemoryResource mallocs pools of memory and uses these to carve out memory for the nodes. Nodes
+ * that are destroyed by the Allocator are actually put back into a free list for further use. This
```
done in e2a5547f12e5da2718ba17ebac5d5aeb5a2e24d5

martinus commented on src/support/allocators/node_allocator.h:
```diff
@@ -0,0 +1,406 @@
+ * in the constructor. Only allocations that match this size will be provided from the preallocated
+ * pools of memory; all other requests simply use ::operator new(). To get the correct node size in
+ * all cases it is unfortunately necessary to have a look at the various implementations of
+ * std::unordered_map. There is currently no standard way of getting the node size programmatically.
```
done in e2a5547f12e5da2718ba17ebac5d5aeb5a2e24d5

martinus commented on src/support/allocators/node_allocator.h:
```diff
@@ -0,0 +1,406 @@
+    template <typename T>
+    [[nodiscard]] T* Allocate(size_t n)
+    {
+        if constexpr (ALLOCATION_SIZE_BYTES == detail::RequiredAllocationSizeBytes<T>()) {
```
Hm, I don't really want to add a counter for `m_free_allocations` but also don't want to iterate through the freelist. I think it's ok to keep the `MemoryResource::DynamicMemoryUsage` as it is because all that memory is the responsibility of that class. I got rid of the `if constexpr` though in e2a5547f12e5da2718ba17ebac5d5aeb5a2e24d5.

martinus commented on src/support/allocators/node_allocator.h:
```diff
@@ -0,0 +1,406 @@
+
+    //! A single linked list of all data available in the MemoryResource. This list is used for
+    //! allocations of single elements.
+    void* m_free_allocations = nullptr;
```
done in e2a5547f12e5da2718ba17ebac5d5aeb5a2e24d5. I also changed the `m_untouched_memory_iterator` and `m_untouched_memory_end` from `void*` to `char*`.

martinus commented on src/memusage.h:
```diff
@@ -148,23 +154,65 @@ static inline size_t DynamicUsage(const std::shared_ptr<X>& p)
+    {
+        // libstdc++, libc++, and MSVC implement the nodes differently. To get the correct size
+        // with the correct alignment, we can simulate that with accordingly nested std::pairs.
+        using ValueType = std::pair<const Key, V>;
```
Moved comment in e2a5547f12e5da2718ba17ebac5d5aeb5a2e24d5

martinus commented on src/memusage.h:
```diff
@@ -148,23 +154,65 @@ static inline size_t DynamicUsage(const std::shared_ptr<X>& p)
+            return sizeof(std::pair<void*, std::pair<ValueType, size_t>>);
+        }
+#endif
+#endif
```
done in e2a5547f12e5da2718ba17ebac5d5aeb5a2e24d5

martinus commented on src/support/allocators/node_allocator.h:
```diff
@@ -0,0 +1,406 @@
+
+    /**
+     * Swapping two containers with unequal allocators who are *not* propagated is undefined
+     * behavior. Unfortunately this is the default! Obviously, we don't want that.
```
changed in e2a5547f12e5da2718ba17ebac5d5aeb5a2e24d5

martinus commented on src/test/node_allocator_tests.cpp:
```diff
@@ -0,0 +1,424 @@
+                      node_allocator::MemoryResourceTester::NumFreeAllocations(mr)); \
+    BOOST_CHECK_EQUAL(num_pools, node_allocator::MemoryResourceTester::NumPools(mr));
+
+#define CHECK_IN_RANGE(what, lowerInclusive, upperInclusive) \
```
done in e2a5547f12e5da2718ba17ebac5d5aeb5a2e24d5

martinus commented on src/test/node_allocator_tests.cpp:
```diff
@@ -0,0 +1,424 @@
+
+BOOST_FIXTURE_TEST_SUITE(node_allocator_tests, BasicTestingSetup)
+
+#define CHECK_MEMORY_RESOURCE(mr, num_free_allocations, num_pools)                   \
```
done in e2a5547f12e5da2718ba17ebac5d5aeb5a2e24d5

martinus commented on src/test/node_allocator_tests.cpp:
```diff
@@ -0,0 +1,424 @@
+        }
+
+        BOOST_CHECK(node_allocator::MemoryResourceTester::NumFreeAllocations(mr) >
+                    num_free_allocations);
```
stricter check in e2a5547f12e5da2718ba17ebac5d5aeb5a2e24d5

martinus commented on src/test/node_allocator_tests.cpp:
```diff
@@ -0,0 +1,424 @@
+template <size_t S>
+struct alignas(32) A32 {
+    char data[S];
+};
```
done in e2a5547f12e5da2718ba17ebac5d5aeb5a2e24d5, and renamed the struct to `AlignedSize`

martinus commented on src/test/node_allocator_tests.cpp:
```diff
@@ -0,0 +1,424 @@
+    }
+
+    // makes sure clear frees all allocations. With MSVC there might be an additional allocation
+    // used for a control structure.
```
Moved that comment up in e2a5547f12e5da2718ba17ebac5d5aeb5a2e24d5

martinus commented on src/test/node_allocator_tests.cpp:
```diff
@@ -0,0 +1,424 @@
+    static_assert(16U == node_allocator::detail::RequiredAllocationSizeBytes<A16<1>>());
+    static_assert(16U == node_allocator::detail::RequiredAllocationSizeBytes<A16<8>>());
+    static_assert(16U == node_allocator::detail::RequiredAllocationSizeBytes<A16<16>>());
+    static_assert(32U == node_allocator::detail::RequiredAllocationSizeBytes<A16<17>>());
```
deduplicated in e2a5547f12e5da2718ba17ebac5d5aeb5a2e24d5. Also I have added a

```cpp
static_assert(__STDCPP_DEFAULT_NEW_ALIGNMENT__ >= std::alignment_of_v<T>, "make sure AllocatePool() aligns correctly");
```
Into `Allocate`, because actually the 16 byte alignments are not allowed when the pools are not correctly aligned as well.

martinus commented on src/test/node_allocator_tests.cpp:
```diff
@@ -0,0 +1,424 @@
+}
+
+
+BOOST_AUTO_TEST_CASE(different_memoryresource_swap)
```
deduplicated in e2a5547f12e5da2718ba17ebac5d5aeb5a2e24d5

sipa commented:
A few high-level comments:
* I think it would be nice to split up the first commit more (e.g. separating the introduction of the allocator from the changes to make the UTXO set use it)
* Don't @ mention people in the PR descriptor; they'll get random spam anytime j-random-altcoin cherry-picks it.
* I find the "semantic" cyclic dependency between node_allocator and memusage (as in: neither can be used without the other, effectively) somewhat ugly. Would it make sense to have the logic for determining node size for containers live in node_allocator itself?

martinus commented:
I have rebased & squashed to e2a5547f12e5da2718ba17ebac5d5aeb5a2e24d5 and integrated most suggestions from @ryanofsky. This update is mostly style improvements, deduplicating code, but also two correctness improvements:

* `MemoryResource::Deallocate` previously just `static_cast` void* to FreeList. This was technically undefined behavior. Instead, I'm now using placement new.
* `MemoryResource::Allocate` now has a `static_assert` to make sure the pool's data is correctly aligned for node alignments.

martinus commented:
@sipa I'll split up the node_allocator into multiple files, and also into multiple commits. It's a bit tricky though to get rid of the cyclic dependency, because MemoryResource needs some functionality of memusage to calculate its dynamic usage, and memusage needs to know the Allocator.

ryanofsky commented:
> It's a bit tricky though to get rid of the cyclic dependency, because MemoryResource needs some functionality of memusage to calculate its dynamic usage, and memusage needs to know the Allocator.

FWIW, I gave a suggestion to remove the circular dependency in https://github.com/bitcoin/bitcoin/pull/22702#discussion_r737676396. Instead of not including the size of allocated nodes in the container size when an allocator is used, the container can report its size the same way whether or not an allocator is used, and the allocator can only report the size of unallocated space instead of the total allocated + unallocated space. I think this would make the accounting more intuitive, in addition to making the container size calculation independent of the allocator size calculation.

martinus commented on src/memusage.h:
```diff
@@ -148,23 +154,65 @@ static inline size_t DynamicUsage(const std::shared_ptr<X>& p)
+    if constexpr (std::is_same_v<Alloc, node_allocator::Allocator<std::pair<const Key, Value>, memusage::NodeSize<ContainerType>::Value()>>) {
+        // Assumes that DynamicUsage of the MemoryResource is called separately. We don't do it here
+        // because multiple maps could use the same MemoryResource.
+        return MallocUsage(sizeof(void*) * m.bucket_count());
```
I've played with this for a while now, but it doesn't work that way because in memusage.h I still need to know if the `node_allocator` is used for a container, because the node sizes are different. For the standard `std::unordered_map` I need to use `memusage::MallocUsage` for the nodes, and for `node_allocator` I have use the node size directly. So I need at least a forward declaration of `node_allocator::Allocator`. But that is fine I guess, I'll create a forward declaration then there's no circular dependency any more.

Also I'll split up `NodeSize` calculation into a separate independent file that both `memusage` and the allocator can use, so there are no circular dependency for that too.

martinus committed:
Make use of node_allocator in CCoinsMap

In my benchmarks, using this pool allocator for CCoinsMap gives about
20% faster `-reindex-chainstate` with -dbcache=5000 with practically the
same memory usage. The change in max RSS changed was 0.3%.

Note that the memory usage behavior is now different, since MemoryResource
does not free any memory when the map is cleared. Instead, the memory is
held in a freelist. This requires now that `CCoinsViewCache::Flush`
needs to call `ReallocateCache` so memory is actually freed. If that
wouldn't happen, memory usage would not decrease and Flush() would be
triggered again soon after.

Also, the `validation_flush_tests` tests need to be updated because
memory allocation is now done in large pools instead of one node at a
time, so the limits need to be updated accordingly.

martinus committed:
Correctly calculate node size for std::unordered_map

Node sizes are calculated correctly for libc++, libstdc++ and MSVC
implementations of std::unordered_map. This is tested for all kind of
possible types used for std::unordered_map.

martinus committed:
Make use of node_allocator::NodeSize in DynamicUsage

Makes use of the new NodeSize calculation in memusage::DynamicUsage
so that the memory usage is more accurate.

martinus commented:
Rebased e2a5547f12e5da2718ba17ebac5d5aeb5a2e24d5 ->  290649c with plenty of refactoring. No functional changes.

* Split code into multiple files, each with its own concern.
* Split up commits into logical chunks
* Added another test for `NodeSize` calculation

martinus commented:
Rebased to 8dfd71ca7fddf1a9436d04a16f5a8d9d98ee8c74 to resolve conflicts

ryanofsky commented:
I'm starting to catch up on this, but so far the new changes look great, and it's nice the way this is now split up into smaller commits and files. I think being able to look at different parts in isolation now makes this PR more approachable and easier to review.

sipa commented on src/support/allocators/node_allocator/node_size.h:
```diff
@@ -0,0 +1,60 @@
+                                                 std::pair<void*, std::pair<ValueType, size_t>>>; // hash stored along ValueType, and that is wrapped with the pointer.
+#endif
+
+    [[nodiscard]] static constexpr size_t Value()
```
Nit: I think with C++17, this can be a constexpr static member variable, rather than a function (`static constexpr size_t VALUE = sizeof(SimulatedNodeType);`).

ryanofsky commented on src/memusage.h:
```diff
@@ -169,6 +170,14 @@ static inline size_t DynamicUsage(const std::unordered_map<Key, Value, Hash, Equ
+template <typename Key, typename Value, typename Hash, typename Equals, typename AllocT, size_t AllocS>
+static inline size_t DynamicUsage(const std::unordered_map<Key, Value, Hash, Equals, node_allocator::Allocator<AllocT, AllocS>>& m)
+{
+    // Since node_allocator is used, the memory of the nodes are the MemoryResource's responsibility and
+    // won't be added here. Also multiple maps could use the same MemoryResource.
+    return MallocUsage(sizeof(void*) * m.bucket_count());
```
In commit "Add allocator for node based containers" (96d8e33acaa079491403d8e8afa6709978dec205)

I still think it's unfortunate to be adding this `memusage` specialization when it is not necessary. I don't like how it:

- Makes memusage.h depend node_allocator, when the job of memusage is just to tell us about memory usage of standard types, and it shouldn't need to be referencing a particular specialized allocator.
- Redefines the concept of memory usage for this one unordered_map+node_allocator container to not consider memory allocated by the container to be "used" by the container, even though the container allocated that memory, and is still responsible for deallocating (if not freeing) it.
- Will require new specializations of map+node_allocator or list+node_allocator to be added here if node_allocator is used with other map or list containers in the future.

I think code can be simpler, definition of "usage" can be more consistent, and cyclic dependencies and special cases can be removed using earlier suggestions from https://github.com/bitcoin/bitcoin/pull/22702#discussion_r737676396, and only downside would be adding a counter variable. On top of the current PR, changes would look like:

```diff
diff --git a/src/Makefile.am b/src/Makefile.am
index b39d03fe8cb..9d09682e243 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -217,7 +217,6 @@ BITCOIN_CORE_H = \
   shutdown.h \
   signet.h \
   streams.h \
-  support/allocators/node_allocator/allocator_fwd.h \
   support/allocators/node_allocator/allocator.h \
   support/allocators/node_allocator/factory.h \
   support/allocators/node_allocator/memory_resource.h \
@@ -575,7 +574,6 @@ libbitcoin_util_a_CPPFLAGS = $(AM_CPPFLAGS) $(BITCOIN_INCLUDES)
 libbitcoin_util_a_CXXFLAGS = $(AM_CXXFLAGS) $(PIE_FLAGS)
 libbitcoin_util_a_SOURCES = \
   support/lockedpool.cpp \
-  support/allocators/node_allocator/memory_resource.cpp \
   chainparamsbase.cpp \
   clientversion.cpp \
   compat/glibcxx_sanity.cpp \
diff --git a/src/memusage.h b/src/memusage.h
index bc20ee4052c..b9faff9b2d9 100644
--- a/src/memusage.h
+++ b/src/memusage.h
@@ -7,7 +7,6 @@
 
 #include <indirectmap.h>
 #include <prevector.h>
-#include <support/allocators/node_allocator/allocator_fwd.h>
 #include <support/allocators/node_allocator/node_size.h>
 
 #include <stdlib.h>
@@ -163,21 +162,13 @@ static inline size_t DynamicUsage(const std::unordered_set<X, Y>& s)
     return MallocUsage(sizeof(unordered_node<X>)) * s.size() + MallocUsage(sizeof(void*) * s.bucket_count());
 }
 
-template <typename Key, typename Value, typename Hash, typename Equals>
-static inline size_t DynamicUsage(const std::unordered_map<Key, Value, Hash, Equals>& m)
+template <typename Key, typename Value, typename Hash, typename Equals, typename Alloc>
+static inline size_t DynamicUsage(const std::unordered_map<Key, Value, Hash, Equals, Alloc>& m)
 {
-    auto node_size = node_allocator::NodeSize<std::unordered_map<Key, Value, Hash, Equals>>::Value();
+    auto node_size = node_allocator::NodeSize<std::unordered_map<Key, Value, Hash, Equals, Alloc>>::Value();
     return MallocUsage(node_size) * m.size() + MallocUsage(sizeof(void*) * m.bucket_count());
 }
 
-template <typename Key, typename Value, typename Hash, typename Equals, typename AllocT, size_t AllocS>
-static inline size_t DynamicUsage(const std::unordered_map<Key, Value, Hash, Equals, node_allocator::Allocator<AllocT, AllocS>>& m)
-{
-    // Since node_allocator is used, the memory of the nodes are the MemoryResource's responsibility and
-    // won't be added here. Also multiple maps could use the same MemoryResource.
-    return MallocUsage(sizeof(void*) * m.bucket_count());
-}
-
 }
 
 #endif // BITCOIN_MEMUSAGE_H
diff --git a/src/support/allocators/node_allocator/allocator.h b/src/support/allocators/node_allocator/allocator.h
index 778471bb20b..c728ab8638a 100644
--- a/src/support/allocators/node_allocator/allocator.h
+++ b/src/support/allocators/node_allocator/allocator.h
@@ -5,7 +5,6 @@
 #ifndef BITCOIN_SUPPORT_ALLOCATORS_NODE_ALLOCATOR_ALLOCATOR_H
 #define BITCOIN_SUPPORT_ALLOCATORS_NODE_ALLOCATOR_ALLOCATOR_H
 
-#include <support/allocators/node_allocator/allocator_fwd.h>
 #include <support/allocators/node_allocator/memory_resource.h>
 
 #include <cstdint>
diff --git a/src/support/allocators/node_allocator/allocator_fwd.h b/src/support/allocators/node_allocator/allocator_fwd.h
deleted file mode 100644
index e3fa0ddd889..00000000000
--- a/src/support/allocators/node_allocator/allocator_fwd.h
+++ /dev/null
@@ -1,17 +0,0 @@
-// Copyright (c) 2021 The Bitcoin Core developers
-// Distributed under the MIT software license, see the accompanying
-// file COPYING or http://www.opensource.org/licenses/mit-license.php.
-
-#ifndef BITCOIN_SUPPORT_ALLOCATORS_NODE_ALLOCATOR_ALLOCATOR_FWD_H
-#define BITCOIN_SUPPORT_ALLOCATORS_NODE_ALLOCATOR_ALLOCATOR_FWD_H
-
-#include <cstddef>
-
-namespace node_allocator {
-
-template <typename T, size_t ALLOCATION_SIZE_BYTES>
-class Allocator;
-
-} // namespace node_allocator
-
-#endif // BITCOIN_SUPPORT_ALLOCATORS_NODE_ALLOCATOR_ALLOCATOR_FWD_H
diff --git a/src/support/allocators/node_allocator/memory_resource.cpp b/src/support/allocators/node_allocator/memory_resource.cpp
deleted file mode 100644
index 368752fa2b9..00000000000
--- a/src/support/allocators/node_allocator/memory_resource.cpp
+++ /dev/null
@@ -1,12 +0,0 @@
-#include <support/allocators/node_allocator/memory_resource.h>
-
-#include <memusage.h>
-
-namespace node_allocator::detail {
-
-size_t DynamicMemoryUsage(size_t pool_size_bytes, std::vector<std::unique_ptr<char[]>> const& allocated_pools)
-{
-    return memusage::MallocUsage(pool_size_bytes) * allocated_pools.size() + memusage::DynamicUsage(allocated_pools);
-}
-
-} // namespace node_allocator::detail
diff --git a/src/support/allocators/node_allocator/memory_resource.h b/src/support/allocators/node_allocator/memory_resource.h
index f4ba0138924..e60e847c86d 100644
--- a/src/support/allocators/node_allocator/memory_resource.h
+++ b/src/support/allocators/node_allocator/memory_resource.h
@@ -8,6 +8,8 @@
 #include <utility>
 #include <vector>
 
+#include <memusage.h>
+
 #ifndef BITCOIN_SUPPORT_ALLOCATORS_NODE_ALLOCATOR_MEMORY_RESOURCE_H
 #define BITCOIN_SUPPORT_ALLOCATORS_NODE_ALLOCATOR_MEMORY_RESOURCE_H
 
@@ -38,12 +40,6 @@ template <typename T>
     return ((size_max + alignment_max - 1U) / alignment_max) * alignment_max;
 }
 
-/**
- * Calculates dynamic memory usage of a MemoryResource. Not a member of MemoryResource so we can implement this
- * in a cpp and don't depend on memusage.h in the header.
- */
-[[nodiscard]] size_t DynamicMemoryUsage(size_t pool_size_bytes, std::vector<std::unique_ptr<char[]>> const& allocated_pools);
-
 } // namespace detail
 
 /**
@@ -102,6 +98,7 @@ public:
     {
         static_assert(ALLOCATION_SIZE_BYTES == detail::RequiredAllocationSizeBytes<T>());
         static_assert(__STDCPP_DEFAULT_NEW_ALIGNMENT__ >= std::alignment_of_v<T>, "make sure AllocatePool() aligns correctly");
+        ++m_allocations;
 
         if (m_free_allocations) {
             // we've already got data in the free list, unlink one element
@@ -138,6 +135,7 @@ public:
     void Deallocate(void* p) noexcept
     {
         static_assert(ALLOCATION_SIZE_BYTES == detail::RequiredAllocationSizeBytes<T>());
+        --m_allocations;
 
         // put it into the linked list.
         //
@@ -150,11 +148,11 @@ public:
     }
 
     /**
-     * Calculates bytes allocated by the memory resource.
+     * Calculates number of bytes allocated to the memory resource minus number of bytes allocated from the memory resource.
      */
     [[nodiscard]] size_t DynamicMemoryUsage() const noexcept
     {
-        return detail::DynamicMemoryUsage(POOL_SIZE_BYTES, m_allocated_pools);
+        return memusage::MallocUsage(POOL_SIZE_BYTES) * m_allocated_pools.size() + memusage::DynamicUsage(m_allocated_pools) - ALLOCATION_SIZE_BYTES * m_allocations;
     }
 
 private:
@@ -171,6 +169,9 @@ private:
         m_untouched_memory_end = m_untouched_memory_iterator + POOL_SIZE_BYTES;
     }
 
+    //! Number of current allocations.
+    size_t m_allocations = 0;
+
     //! Contains all allocated pools of memory, used to free the data in the destructor.
     std::vector<std::unique_ptr<char[]>> m_allocated_pools{};
 

```

martinus replied:
Thanks for the patch! This makes discussion a lot easier. I think there's one bug in it, this now overestimates memory usage. In `memusage.h` all maps are now calculated like this:

```cpp
return MallocUsage(node_size) * m.size() + MallocUsage(sizeof(void*) * m.bucket_count());
```
To correct for the incorrectly added `MallocUsage(node_size) * m.size()` it should be `memoryusage::MallocUsage(ALLOCATION_SIZE_BYTES) * m_allocations`. That should be correct in most cases, but it might not be when `RequiredAllocationSizeBytes<typename NodeSize...>` (in `factory.h`) gives a different result than the NodeSize. The only way to do this correctly is to also pass the result of `NodeSize` to the `MemoryResource`, but this complicates things.

I also don't like that in most cases `memoryresource.DynamicMemoryUsage()` will return a negative number (causing `size_t` underflow) to correct for the overestimation in memusage. This will happen whenever the overestimation is large enough.

martinus replied:
Here is a concrete example that I just tried with `CCoinsMap`. I created a CCoinsMap and added 100k entries. `memusage::DynamicUsage` calculates this:

```cpp
return MallocUsage(node_size) * m.size() + MallocUsage(sizeof(void*) * m.bucket_count());
// return MallocUsage(104) * 100000 + MallocUsage(8 * 172933)  
// return 128 * 100000 + 1383488
// return 14183488
```

So `DynamicUsage` estimates 13.5MB memory usage. This is too much, because `node_allocator` doesn't have the overhead of malloc. `MallocUsage` uses 128 bytes per node, but when `node_allocator` the bytes per node are actually just 104. So when calculating `MemoryResource::DynamicMemoryUsage` we'd have to subtract the 128*100000 and not 104*100000 to get the correct memory usage:

```cpp
return memusage::MallocUsage(POOL_SIZE_BYTES) * m_allocated_pools.size() + memusage::DynamicUsage(m_allocated_pools) - ALLOCATION_SIZE_BYTES * m_allocations;
// return MallocUsage(262080) * 40 + 528 - 104 * 100000 // <-- incorrect, has to be 128*100000
// return 262096 * 40 + 528 - 128*100000 // correct, but no easy way to get the correct 128 bytes within MemoryResource
// return -2315632
```

ryanofsky reviewed (approved):
Code review ACK 8dfd71ca7fddf1a9436d04a16f5a8d9d98ee8c74. Finally caught up on this again! Appreciate the new count-allocations test, and all the breakups and cleanups.

I did leave some more suggestions below about not making memusage.h depend on node_allocator, and only making the dependency go the other way. But I raised concerns about that before, and think they are minor, and it's fine ignore them and stick with the current approach. Overall everything looks very good here!

jonatack commented:
Starting review to catch up on the changes since my previous review at 952c37a31c39b7acdb5ab2634a354a1c548708a2.

martinus commented on src/support/allocators/node_allocator/memory_resource.h:
```diff
@@ -0,0 +1,196 @@
+    [[nodiscard]] T* Allocate()
+    {
+        static_assert(ALLOCATION_SIZE_BYTES == detail::RequiredAllocationSizeBytes<T>());
+        static_assert(__STDCPP_DEFAULT_NEW_ALIGNMENT__ >= std::alignment_of_v<T>, "make sure AllocatePool() aligns correctly");
```
I think that check is necessary. In `AllocatePool` I'm doing `new char[POOL_SIZE_BYTES]`, and this uses `__STDCPP_DEFAULT_NEW_ALIGNMENT__` as the alignment for the pool's memory. So if you try to use the node_allocator with a type that has an alignment > `__STDCPP_DEFAULT_NEW_ALIGNMENT__` this is a problem, even when `detail::RequiredAllocationSizeBytes` is done correctly. I had a test case that used 16 byte alignment, and with that `static_assert` that test failed on a 32bit machine.

martinus replied:
I think I'd prefer to keep the `static_assert` at the `Allocate` entry point, because currently `AllocatePool` doesn't need to know the type `T`. I guess it's more a matter of style :)

martinus commented on src/test/node_allocator_helpers.h:
```diff
@@ -0,0 +1,27 @@
+ */
+template <typename T>
+struct NotNoexceptHash {
+    size_t operator()(const T& x) const /* not noexcept */
```
I think this is safe, the standard says that only a handful functions are implicitly nothrowing: default ctor and dtor, move & copy ctor, move & assignment operator. There has been a proposal for a `noexcept(auto)` to add implicitly nothrowing anywhere else, but it seems that there were too many problems.

martinus commented on src/test/node_size_tests.cpp:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    // there should be a entry with exactly 123 allocations
+    auto it = std::find_if(SingletonAllocationInfo().begin(), SingletonAllocationInfo().end(), [](const auto& kv) {
```
This is unfortunately not the case on Windows, there the allocator is used to allocate one control structure as well

martinus commented:
Rebased to 02da536 to resolve conflicts, and some very minor changes:
* Fixed some comments
* use `static constexpr size_t VALUE` in `NodeSize`

martinus committed:
Add allocator for node based containers

This allocator is potentially useful for all node-based containers like
`std::list`, `std::map`, `std::unordered_map` that are heavily used.

The allocator uses pools to be more memory efficient and faster than
the default std::allocator. Currently this provides specializations for
`std::unordered_map`, and other specializations can be added as needed.

This updates the `memusage::DynamicUsage` calculation to take the
node_allocator allocator into account.

Also, the dynamic usage for standard `std::unordered_map` is made more
accurate with the new node_size calculation.

Rule of 5, drop unused function, spelling fixups and more code cleanup
by jonatack, countless suggestions & improvements by ryanofsky.

Co-authored-by: Jon Atack <jon@atack.com>
Co-authored-by: Russell Yanofsky <russ@yanofsky.org>

martinus commented:
Rebased to resolve conflict in src/Makefile.bench.include

jonatack commented:
Now that v23 has been branched off, this would be a good time to move forward on this. Planning to re-review.

laanwj commented:
> Now that v23 has been branched off, this would be a good time to move forward on this. 

Agree, it would be good to get this in early in the release cycle. I'll start testing it as well.

martinus commented:
Rebased to resolve conflict in `Makefile.test.include`

jonatack commented:
> Rebased to resolve conflict in `Makefile.test.include` By the way, is there an easy way to see what has actually changed in a rebase? I just moved 2 lines.

`git range-diff a19f641 472f5b0 95c53bf`

(Will re-review this PR soon.)

laanwj commented:
I have been testing this on a busy node (~400 connections) for more than a month. I have noticed no crashes, no memory leaks, and no other unexpected behavior.

sipa commented:
I see a number of unaddressed comments here by @ryanofsky. What is the status of those?

martinus commented:
> I see a number of unaddressed comments here by @ryanofsky. What is the status of those?

I've now gone through all unaddressed comments and marked all as resolved that I consider fully resolved. As far as I can say only nits remain

laanwj commented:
> I have been testing this on a busy node (~400 connections) for more than a month. I have noticed no crashes, no memory leaks, and no other unexpected behavior.

Still no issues! 

Still, to be honest, I think this change is a bit scary, and I have a hard time signing off on it. The reason is that it adds quite a lot of complexity, and depends on quite some C++ magic and cleverness (e.g. relying on `unordered_map` internals). C++ being not exactly a memory-safe language, this increases the risk of accidental foot-guns. I appreciate the work it's great to make things faster, and I really don't want to discourage it, but, especially as long as it's used for consensus code we do have to err on the side of safety and correctness.

martinus commented:
> Still, to be honest, I think this change is a bit scary

I guess it is, it got a lot more complex that I initially anticipated. I don't think it can be made much simpler. I'm pretty sure it works, but obviously there's no 100% guarantee. So I'm fine with it not being merged, at least then I can't be responsible for a bitcoind crash :sweat_smile: 

Since the code here is not tightly coupled to bitcoind, when I have the time I will try to create a generic c++ library out of it.

Thanks everyone who reviewed & tested or otherwise spent time on this!

jonatack commented:
@martinus this pull was discussed yesterday at the weekly meeting (see https://www.erisian.com.au/bitcoin-core-dev/log-2022-06-02.html#l-358). I've been running it without issues as well for some time and am concept ACK (but need to re-review carefully). @sipa wrote "I need to look at the latest code changes again, but I'm very concept ACK on that allocator". For info :)

jamesob commented:
> I'm trying to understand what potential problems we fear here. Is it that the node crashes or is it consensus problems (or both)?

I think the major concern here could be summarized as "we're worried about unknown unknowns." Swapping out an allocator (or as in the original case of #16718, the whole map implementation) entails displacing a lot of low-level C++ code that is widely deployed and tested by the standard library with code that we (or if not just us, a much smaller subset of the software world) maintain and test. 

If anything goes wrong with the UTXO set, it's no news to everyone that it certainly spells big issues for bitcoin. While the performance gains here are very enticing, and indeed @luke-jr earlier made the interesting point that maybe moving this critical code into our domain of control (and out of the black box of std) may be preferable, since we would explicitly understand and maintain a very core dependency of the validation process, this "mountain man" approach obviously has an associated cost: maintaining such an important and sensitive piece of the system represents a liability that we would no longer shoulder alongside the wider C++ usage base. 

I originally proposed the hashmap implementation change that (I think?) inspired @martinus to undertake this work. I'm glad he has done this week, even if it remains experimental, because it's important for us to understand this design space better. But that said I can understand why @laanwj would express an abundance of caution about this kind of thing, and it resonates with me. Everyone here understands that a change like this should not be done lightly, but even given that, it's hard to know when we've "tested enough" to know that this will be safe for perpetuity. Unknown unknowns. 

All that said, I'm far from the most qualified systems guy on this project, so it would be interesting to hear from others like @sipa @ryanofsky @theuni @TheBlueMatt. Maybe we can fuzz this to death on a bunch of different platforms and convince ourselves that it's as safe or safer than using std's allocator. Such a determination is out of my ability.

martinus commented:
I think I've found a way to get almost all of the benefit of the `node_allocator` with a significantly less scary implementation. Same memory advantage, but a bit slower. I'll prepare a new PR :+1:

sipa commented:
@martinus I started working on a fuzz test for the allocator, to show at least its behavior w.r.t. its external interface is correct.

If you're going to simplify things: will there still be an allocator class, and a concept of a memory resource? If so, will the memory resource still be parametrized by the size of the pooled objects in it?

martinus commented:
> @martinus I started working on a fuzz test for the allocator, to show at least its behavior w.r.t. its external interface is correct.

That's awesome!

> If you're going to simplify things: will there still be an allocator class, and a concept of a memory resource? If so, will the memory resource still be parametrized by the size of the pooled objects in it?

There will be just a memory resource, which will be passed to a `std::pmr::polymorphic_allocator`. I have a prototype of the memory resource here: https://gist.github.com/martinus/0cee4dd177beda78bb1e2e3d128c2d70
That's basically all that is needed in <100 lines of code. This implements the abstract `std::pmr::memory_resource` interface which is much simpler than implementing a full allocator. Doesn't need any rebinding, type casting, reinterpret_cast.

Also, instead of having to correctly guess the pool's memory size I simply use a vector with multiple pools of different size. Most of them will stay empty, but that overhead is negligible. 

Usage is like so:
```cpp
using CCoinsMap = std::pmr::unordered_map<COutPoint, CCoinsCacheEntry, SaltedOutpointHasher>;
auto mr = NodePoolResource<256>(); // large enough to fit nodes size
auto map = CCoinsMap{0, SaltedOutpointHasher{}, std::equal_to<COutPoint>{}, &mr};
```

martinus commented:
> Another idea for something to include in the tests which I haven't done: add a member function to the resource to ask if it still has any outstanding allocations (by summing the size of all allocated chunks, and comparing with the sum of all elements in the freelist).

Excellent idea, I've now added an even stronger check: make sure that all data in the freelist actually comes from the chunks, doesn't overlap in any way, and each byte from the chunks can be found in the freelists or is still available
